lib/index.js

```diff
@@ -58,88 +56,84 @@ api.use = options => {
-          if(options.reuse && req.isAuthenticated()) {
-            return done(null, req.user);
-          }
-          return passport.authenticate('session', options)(
-            req, res, err => done(err, req.user || false));
+    // restrict session-based authentication to when:
+    // 1. There is no origin header set.
+    // 2. The origin header matches the host header.
+    // 3. The request method is in a list of default permitted CORS methods.
+    // TODO: add an option to allow controlling permittedCorsMethods for
+    // particular handlers that know what they're doing
+    const origin = ('origin' in req.headers ?
+      URL.parse(req.headers.origin).host : null);
+    if(origin === null || req.headers.host === origin ||
+      permittedCorsMethods.indexOf(req.method) !== -1 ||
+      _checkAllowedHosts(origin, options.allowHosts)) {
+      if(options.allowUrlEncoded || !(
+        req.is('urlencoded') || req.is('multipart'))) {
+        // reuse existing check (express auto-runs a session check per
+        // code below with
+        // `app.use(api.createMiddleware({strategy: 'session'}))`
+        // TODO: potentially refactor to avoid this auto-check in the future,
+        // but it would be a breaking change
+        // FIXME: `req.isAuthenticated` is not technically granular enough
+        // to determine if `session` strategy was responsible for previous
+        // authentication, but presently works because session is always
+        // checked and if another method was concurrently checked, both would
+        // fail if the user didn't match; however, should the automatic
+        // session check be removed, then the degenerate case where a user
+        // was authenticated via another method and then checked against
+        // session this will pass when it should potentially fail
+        if(options.reuse && req.isAuthenticated()) {
+          return emit({user: req.user || false});
         }
+        return new Promise((resolve, reject) => {
+          passport.authenticate('session', options)(req, res, err => {
+            if(err) {
+              return reject(err);
+            }
+            resolve(emit({user: req.user || false}));
+          });
+        });
       }
-      return done(null, false);
-    };
-  }
-
-  if(typeof callback !== 'function') {
-    throw new TypeError('callback must be a function.');
+    }
+    return {user: false};
   }
 
   // handle non-built-in strategies
-  return passport.authenticate(strategy, options, emitter(callback));
-
-  function emitter(callback) {
-    return (err, user, info) => {
-      if(!err && user) {
-        return bedrock.events.emit('bedrock-passport.authenticate', {
-          strategy: strategy,
-          options: options,
-          user: user,
-          info: info
-        }, err => callback(err, user, info));
+  return new Promise((resolve, reject) => {
+    passport.authenticate(strategy, options)(req, res, err => {
```

Comment:
not familiar with this syntax, is it OK?

Reply:
Yes... I think so... dunno what you mean. The `passport.authenticate` call returns a middleware that we immediately invoke here.

Reply:
OK, good.